File: app/src/main/java/com/example/bridgeapp/structure/BidObject.java
package com.example.bridgeapp.structure;

import android.os.Parcel;
import android.os.Parcelable;

import com.example.bridgeapp.util.CardSuit;

public class BidObject implements Comparable<BidObject>, Parcelable {

    private final int number;
    private final CardSuit suit;
    private final int doubleVal;
    private final boolean pass;

    public BidObject(int number, String suit) {
        this(number, suit, false);
    }

    public BidObject(int number, String suit, boolean pass) {
        this.number = number;

        System.out.println(suit.trim());
        this.suit = CardSuit.valueOf(suit.trim());

        doubleVal = 0;
        this.pass = pass;
    }

    public BidObject(BidObject o, int doubleVal, boolean pass){
        this.number = o.number;
        this.suit = o.suit;
        this.doubleVal = doubleVal;
        this.pass = pass;
    }

    public BidObject(BidObject o, int doubleVal){
        this(o, doubleVal, false);
    }

    public BidObject(BidObject o, boolean pass){
        this(o, 0, pass);
    }

    public static final Creator<BidObject> CREATOR = new Creator<BidObject>() {
        @Override
        public BidObject createFromParcel(Parcel in) {
            return new BidObject(in);
        }

        @Override
        public BidObject[] newArray(int size) {
            return new BidObject[size];
        }
    };

    public int getNumber() {
        return number;
    }

    public CardSuit getSuit() {
        return suit;
    }

    public int getDoubleVal() {
        return doubleVal;
    }

    public boolean isPass() {
        return pass;
    }

    public int getValue(){
        return this.number * 100 + this.suit.ordinal() * 10 + this.doubleVal;
    }

    @Override
    public int compareTo(BidObject o){
        int thisValue = this.getValue();
        int otherValue = o.getValue();

        if(thisValue - otherValue == 0)
            return (this.pass ? 1 : 0) - ( o.pass ? 1 : 0);
        else
            return thisValue - otherValue;
    }

    @Override
    public String toString(){
        if(pass) return "PASS";

        StringBuilder out = new StringBuilder(String.valueOf(number));
        out.append(suit.symbol);
        for(int i = 0; i < doubleVal; i++)
            out.append("X");

        return out.toString();
    }

    protected BidObject(Parcel in) {
        number = in.readInt();
        suit = CardSuit.valueOf(in.readString());
        doubleVal = in.readInt();
        pass = in.readByte() != 0;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(number);
        dest.writeString(this.suit.name());
        dest.writeInt(doubleVal);
        dest.writeByte((byte) (pass ? 1 : 0));
    }
}

File: app/src/main/java/com/example/bridgeapp/structure/BidObject.kt
package com.example.bridgeapp.structure

import android.os.Parcel
import android.os.Parcelable
import com.example.bridgeapp.util.CardSuit
import java.io.Serializable

class BidObject : Comparable<BidObject>, Parcelable, Serializable {
    val number: Int
    val suit: CardSuit
    val doubleVal: Int
    val isPass: Boolean

    @JvmOverloads
    constructor(number: Int, suit: String, pass: Boolean = false) {
        this.number = number
        println(suit.trim { it <= ' ' })
        this.suit = CardSuit.valueOf(suit.trim { it <= ' ' })
        doubleVal = 0
        isPass = pass
    }

    @JvmOverloads
    constructor(o: BidObject, doubleVal: Int, pass: Boolean = false) {
        number = o.number
        suit = o.suit
        this.doubleVal = doubleVal
        isPass = pass
    }

    constructor(o: BidObject, pass: Boolean) : this(o, 0, pass)

    private val value: Int
        get() = number * 100 + suit.ordinal * 10 + doubleVal

    override fun compareTo(other: BidObject): Int {
        val thisValue = value
        val otherValue = other.value
        return if (thisValue - otherValue == 0) (if (isPass) 1 else 0) - (if (other.isPass) 1 else 0) else thisValue - otherValue
    }

    override fun toString(): String {
        if (isPass) return "PASS"
        val out = StringBuilder(number.toString())
        out.append(suit.symbol)
        for (i in 0 until doubleVal) out.append("X")
        return out.toString()
    }

    private constructor(`in`: Parcel) {
        number = `in`.readInt()
        suit = CardSuit.valueOf(`in`.readString()!!)
        doubleVal = `in`.readInt()
        isPass = `in`.readByte().toInt() != 0
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(number)
        dest.writeString(suit.name)
        dest.writeInt(doubleVal)
        dest.writeByte((if (isPass) 1 else 0).toByte())
    }

    companion object {
        @JvmField val CREATOR: Parcelable.Creator<BidObject> = object : Parcelable.Creator<BidObject> {
            override fun createFromParcel(`in`: Parcel): BidObject {
                return BidObject(`in`)
            }

            override fun newArray(size: Int): Array<BidObject?> {
                return arrayOfNulls(size)
            }
        }
    }
}